c_src/optics_htable/htable.c

```diff
@@ -0,0 +1,230 @@
+/* htable.c
+   Rémi Attab (remi.attab@gmail.com), 10 Mar 2016
+   FreeBSD-style copyright and disclaimer apply
+*/
+
+#include <stdlib.h>
+#include <assert.h>
+#include <limits.h>
+#include "htable.h"
+
+extern inline uint64_t htable_hash(const struct Key *key);
+
+// -----------------------------------------------------------------------------
+// hash
+// -----------------------------------------------------------------------------
+
+// FNV-1a hash implementation: http://isthe.com/chongo/tech/comp/fnv/
+inline uint64_t htable_hash(const struct Key *key)
+{
+    uint64_t hash = 0xcbf29ce484222325;
+    for (size_t i = 0; i < key->n; ++i)
+        hash = (hash ^ key->bytes[i]) * 0x100000001b3;
+
+    return hash;
+}
+
+
+// -----------------------------------------------------------------------------
+// config
+// -----------------------------------------------------------------------------
+
+enum { probe_window = 8 };
+
+
+// -----------------------------------------------------------------------------
+// basics
+// -----------------------------------------------------------------------------
+
+void htable_reset(struct htable *ht)
+{
+    if (ht->table) {
+        for (size_t i = 0; i < ht->cap; ++i) {
+            if (ht->table[i].key) free((struct Key *) ht->table[i].key);
+        }
+        free(ht->table);
+    }
+    *ht = (struct htable) {0};
+}
+
+static bool table_put(
+        struct htable_bucket *table, size_t cap,
+        const struct Key *key, uint64_t value)
+{
+    uint64_t hash = htable_hash(key);
+
+    for (size_t i = 0; i < probe_window; ++i) {
+        struct htable_bucket *bucket = &table[(hash + i) % cap];
+        if (bucket->key) continue;
+
+        bucket->key = key;
+        bucket->value = value;
+        return true;
+    }
+
+    return false;
+}
+
+static void htable_resize(struct htable *ht, size_t cap)
+{
+    if (cap <= ht->cap) return;
+
+    size_t new_cap = ht->cap ? ht->cap : 1;
+    while (new_cap < cap) new_cap *= 2;
+
+    struct htable_bucket *new_table = calloc(new_cap, sizeof(*new_table));
+    assert(new_table != NULL);
+
+    for (size_t i = 0; i < ht->cap; ++i) {
+        struct htable_bucket *bucket = &ht->table[i];
+        if (!bucket->key) continue;
+
+        if (!table_put(new_table, new_cap, bucket->key, bucket->value)) {
+            free(new_table);
+            htable_resize(ht, new_cap * 2);
+            return;
+        }
+    }
+
+    free(ht->table);
+    ht->cap = new_cap;
+    ht->table = new_table;
+}
+
+void htable_reserve(struct htable *ht, size_t items)
+{
+    htable_resize(ht, items * 4);
+}
+
+struct Key *htable_key_int(uint64_t i)
+{
+    struct Key *key = malloc(sizeof(struct Key) + sizeof(int));
+    assert(key != NULL);
+    key->n = sizeof(int);
```

Comment:
your function is accepting an `uint64_t` as a parameter which is not consistent with the body of the function.